cpp/tools/parquet/parquet_dump_footer.cc

```diff
@@ -0,0 +1,248 @@
+// under the License.
+
+#include <cstdint>
+#include <cstring>
+#include <exception>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <optional>
+#include <random>
+
+#include <thrift/protocol/TCompactProtocol.h>
+#include <thrift/transport/TBufferTransports.h>
+
+#include "arrow/filesystem/filesystem.h"
+#include "arrow/util/endian.h"
+#include "arrow/util/ubsan.h"
+#include "generated/parquet_types.h"
+#include "parquet/thrift_internal.h"
+
+using apache::thrift::protocol::TCompactProtocol;
+using apache::thrift::transport::TMemoryBuffer;
+using apache::thrift::transport::TTransport;
+
+namespace {
+int PrintHelp() {
+  std::cerr << R"(
+Usage: parquet-dump-footer
+  -h|--help    Print help and exit
+  --no-scrub   Do not scrub potentially PII metadata
+  --json       Output JSON instead of binary
+  --in         Input file: required
+  --out        Output file: defaults to stdout
+
+  Dumps the footer of a Parquet file to stdout or a file, optionally with
+  potentially user specific metadata scrubbed.
+)";
+  return 1;
+}
+
+uint32_t ReadLE32(const void* p) {
+  uint32_t x = arrow::util::SafeLoadAs<uint32_t>(static_cast<const uint8_t*>(p));
+  return arrow::bit_util::FromLittleEndian(x);
+}
+
+void AppendLE32(uint32_t v, std::string* out) {
+  v = arrow::bit_util::ToLittleEndian(v);
+  out->append(reinterpret_cast<const char*>(&v), sizeof(v));
+}
+
+template <typename T>
+bool Deserialize(const char* data, uint32_t len, T* obj) {
+  parquet::ThriftDeserializer des(/*string_size_limit=*/1 << 30,
+                                  /*container_size_limit=*/1 << 30);
+  try {
+    des.DeserializeMessage(reinterpret_cast<const uint8_t*>(data), &len, obj);
+    return true;
+  } catch (const std::exception& e) {
+    std::cerr << "Failed to deserialize: " << e.what() << "\n";
+    return false;
+  }
+}
+
+template <typename T>
+bool Serialize(const T& obj, std::string* out) {
+  parquet::ThriftSerializer ser(/*initial_buffer_size=*/10 << 20);
+  try {
+    ser.SerializeToString(&obj, out);
+    return true;
+  } catch (const std::exception& e) {
+    std::cerr << "Failed to serialize: " << e.what() << "\n";
+    return false;
+  }
+}
+
+// Replace the contents of s with random data of the same length.
+void Scrub(std::string* s) {
+  static char pool[4096];
+  static std::mt19937 rng(std::random_device{}());
+  static const bool kPoolInit = [] {
+    std::uniform_int_distribution<> caps(65, 90);
+    for (size_t i = 0; i < sizeof(pool); i++) pool[i] = caps(rng);
+    return true;
+  }();
+  (void)kPoolInit;
+
+  const size_t n = s->size();
+  s->clear();
+  while (s->size() < n) {
+    // To avoid repeating patterns we start somewhere up to halfway through the pool and
+    // append up to half the pool chars.
+    std::uniform_int_distribution<size_t> half(0, sizeof(pool) / 2 - 1);
+    size_t m = std::min(n, half(rng) + 1 /* at least one */);
+    s->append(&pool[half(rng)], m);
+  }
```

Comment:
This seems very complicated. Why not simply overwrite `s` with random chars?
Something like (untested):
```suggestion
  std::uniform_int_distribution<unsigned int> char_dist(0, 255);
  for (char& c : *s) {
    c = static_cast<char>(char_dist(rng));
  }
```

Reply:
(if you're worried about RNG speed, you can use one of the PCG engines from `arrow/util/pcg_random.h`)

Reply:
Done.